server/common/EntryParser.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (C) 2021 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.server.common;
+
+import com.vaticle.typedb.common.yaml.Yaml;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.server.common.CommandLine.Option;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.vaticle.typedb.common.util.Objects.className;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_CAST;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.CONFIG_OPTION_ENUM_UNEXPECTED_VALUE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.CONFIG_OPTION_UNEXPECTED_VALUE_TYPE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.MISSING_CONFIGURATION_OPTION;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+import static com.vaticle.typedb.core.server.common.Util.scopeOption;
+
+public abstract class EntryParser implements Option.CliHelp {
```

Comment:
Each entry in the configuration file is specified with an `EntryParser` - these can be `Named.Unrestricted`, `Named.Enum`, or `Unnamed`. Each entry is also paired with a `ValueParser`, which parses the value in a key-value pair. This can be a simple Leaf value, or a Nested value. All the `Configuration` Parsers after the root parser are `ValueParser`'s, which take the value of a key-value section of the YAML configuration, and convert it into an object.